File: Dumplings/src/main/java/net/sinclairstudios/dumplings/ui/fragment/DumplingNameDialogFragment.java
package net.sinclairstudios.dumplings.ui.fragment;

import android.app.AlertDialog;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.FragmentManager;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;
import net.sinclairstudios.dumplings.R;
import net.sinclairstudios.dumplings.domain.HasDumpling;
import net.sinclairstudios.dumplings.ui.binding.DumplingBinderFactory;
import net.sinclairstudios.dumplings.ui.widgets.ListenerTrackingEditText;
import org.jetbrains.annotations.NotNull;

public class DumplingNameDialogFragment extends DialogFragment {

    private final DumplingBinderFactory dumplingBinderFactory;
    private final HasDumpling hasDumpling;

    public DumplingNameDialogFragment(DumplingBinderFactory dumplingBinderFactory,
                                      HasDumpling hasDumpling) {
        this.dumplingBinderFactory = dumplingBinderFactory;
        this.hasDumpling = hasDumpling;
    }

    @NotNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        LayoutInflater inflater = getActivity().getLayoutInflater();

        View dialogView = inflater.inflate(R.layout.dumpling_name_layout, null);

        builder
                .setView(dialogView)
                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(@NotNull DialogInterface dialog, int which) {
                        DumplingNameDialogFragment.this.getDialog().dismiss();
                    }
                });


        ListenerTrackingEditText editText =
                (ListenerTrackingEditText) dialogView.findViewById(R.id.dumplingNameEditText);
        ImageView imageView = (ImageView) dialogView.findViewById(R.id.dumplingImage);

        dumplingBinderFactory.bindListenerTrackingEditText(hasDumpling, editText);
        dumplingBinderFactory.bindImageView(editText, imageView);
        return builder.create();
    }

    @Override
    public void onResume() {
        super.onResume();
        getDialog().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE);
    }

    public static class Spawner implements View.OnFocusChangeListener {

        private final FragmentManager parent;
        private final DumplingBinderFactory dumplingBinderFactory;
        private final HasDumpling hasDumpling;

        private DumplingNameDialogFragment dialogFragment;

        public Spawner(FragmentManager parent, DumplingBinderFactory dumplingBinderFactory,
                       HasDumpling hasDumpling) {
            this.parent = parent;
            this.dumplingBinderFactory = dumplingBinderFactory;
            this.hasDumpling = hasDumpling;
        }

        @Override
        public void onFocusChange(View v, boolean hasFocus) {
            if (hasFocus) {
                if (dialogFragment == null) {
                    dialogFragment = new DumplingNameDialogFragment(dumplingBinderFactory, hasDumpling);
                    dialogFragment.show(parent, "DumplingNameDialogFragment-" + hasDumpling.toString());
                }
            }
        }
    }
}

File: app/src/main/java/net/sinclairstudios/dumplings/activity/DumplingNameDialogFragment.kt
package net.sinclairstudios.dumplings.activity

import android.app.AlertDialog
import android.app.Dialog
import android.os.Bundle
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import net.sinclairstudios.dumplings.R
import net.sinclairstudios.dumplings.binding.DumplingBinderFactory
import net.sinclairstudios.dumplings.domain.HasDumpling
import net.sinclairstudios.dumplings.widget.ListenerTrackingEditTextView

class DumplingNameDialogFragment(val dumplingBinderFactory: DumplingBinderFactory, val hasDumpling: HasDumpling) : DialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(activity)
        val inflater = requireActivity().layoutInflater

        val dialogView: View = inflater.inflate(R.layout.dumpling_name_layout, null)

        builder
            .setView(dialogView)
            .setPositiveButton(
                android.R.string.ok
            ) { dialog, which -> this@DumplingNameDialogFragment.dialog!!.dismiss() }


        val editText: ListenerTrackingEditTextView = dialogView.findViewById(R.id.dumplingNameEditText)
        val imageView = dialogView.findViewById<View>(R.id.dumplingImage) as ImageView

        dumplingBinderFactory.bindListenerTrackingEditText(hasDumpling, editText)
        dumplingBinderFactory.bindImageView(editText, imageView)
        return builder.create()
    }

    override fun onResume() {
        super.onResume()
        requireDialog().window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_STATE_ALWAYS_VISIBLE)
    }

    class Factory(val parent: FragmentManager,
                        val dumplingBinderFactory: DumplingBinderFactory,
                        val hasDumpling: HasDumpling) : View.OnClickListener {

        private var dialogFragment: DumplingNameDialogFragment? = null

        override fun onClick(view: View?) {
            if (dialogFragment == null) {
                dialogFragment = DumplingNameDialogFragment(dumplingBinderFactory, hasDumpling).also {
                    it.show(parent, "DumplingNameDialogFragment-$hasDumpling")
                }
            }
        }
    }
}